projects/driver_controls/src/pedal_calibration.c

```diff
@@ -0,0 +1,192 @@
+      break;
+    }
+  }
+
+  // Check if there are enough samples taken for the channel.
+  if (storage->sample_counter[channel] >= PEDAL_CALIBRATION_NUM_SAMPLES) {
+    // If yes, stop sampling and remove the callback.
+    ads1015_configure_channel(storage->settings.ads1015_storage, ads1015_channel, false, NULL,
+                              NULL);
+  } else {
+    // If not, continue reading.
+    ads1015_read_raw(storage->settings.ads1015_storage, ads1015_channel, &reading);
+    storage->sample_counter[channel]++;
+
+    PedalCalibrationRange *range = &storage->band[channel][state];
+    range->min = MIN(range->min, reading);
+    range->max = MAX(range->max, reading);
+  }
+}
+
+// Initializes the calibration storage. Requires initialized Ads1015storage.
+StatusCode pedal_calibration_init(PedalCalibrationStorage *storage,
+                                  PedalCalibrationSettings *settings) {
+  // TODO(ELEC-382): validate inputs
+  memset(storage, 0, sizeof(*storage));
+  storage->settings = *settings;
+
+  return STATUS_CODE_OK;
+}
+
+// Given the state (full throttle/brake), it finds the min and max raw readings of each channel.
+// These points would become the vertices of the band we want to obtain around the data.
+StatusCode pedal_calibration_process_state(PedalCalibrationStorage *storage,
+                                           PedalCalibrationState state) {
+  if (storage == NULL || state >= NUM_PEDAL_CALIBRATION_STATES) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  bool disabled = critical_section_start();
+  storage->state = state;
+  storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_A] = 0;
+  storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_B] = 0;
+
+  int16_t reading;
+  for (PedalCalibrationChannel channel = PEDAL_CALIBRATION_CHANNEL_A;
+       channel < NUM_PEDAL_CALIBRATION_CHANNELS; channel++) {
+    storage->band[channel][state].min = INT16_MAX;
+    storage->band[channel][state].max = INT16_MIN;
+  }
+  critical_section_end(disabled);
+
+  // Set up callback on channels which would read new values after each conversion.
+  // Since we disable the channel reads as we obtain enough samples, they need to be reconfigured
+  // for the new stage.
+  status_ok_or_return(ads1015_configure_channel(
+      storage->settings.ads1015_storage, storage->settings.adc_channel[PEDAL_CALIBRATION_CHANNEL_A],
+      true, prv_adc_callback, storage));
+  status_ok_or_return(ads1015_configure_channel(
+      storage->settings.ads1015_storage, storage->settings.adc_channel[PEDAL_CALIBRATION_CHANNEL_B],
+      true, prv_adc_callback, storage));
+
+  // Wait until both channels finish sampling.
+  while (storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_A] < PEDAL_CALIBRATION_NUM_SAMPLES &&
+         storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_B] < PEDAL_CALIBRATION_NUM_SAMPLES) {
+  }
+
+  return STATUS_CODE_OK;
+}
+
+static void prv_calc_midline(PedalCalibrationStorage *storage,
```

Comment:
Nit: `s/midline/midpoint/`

Reply:
It's not really the midpoint? It's creating a line from two midpoints I suppose.